File: src-tauri/src/dap.rs
use serde::{Deserialize, Serialize};
use std::env;
use std::io::{BufRead, BufReader, Read, Write};
use std::path::{Path, PathBuf};
use std::process::{Command, Stdio};
use std::sync::{Arc, Mutex};
use std::thread;
use tauri::{AppHandle, Emitter, Runtime};

pub struct DapState {
    pub child: Arc<Mutex<Option<std::process::Child>>>,
    pub stdin: Arc<Mutex<Option<std::process::ChildStdin>>>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct DapCommandCandidate {
    pub cmd: String,
    pub args: Vec<String>,
}

#[derive(Debug, Clone, Serialize)]
pub struct ResolvedDapCommand {
    pub cmd: String,
    pub args: Vec<String>,
    pub source: String,
}

#[derive(Debug, Clone, Deserialize)]
pub struct DapAdapterProbe {
    pub id: String,
    pub label: String,
    pub languages: Vec<String>,
    pub candidates: Vec<DapCommandCandidate>,
}

#[derive(Debug, Clone, Serialize)]
pub struct DapAdapterStatus {
    pub id: String,
    pub label: String,
    pub languages: Vec<String>,
    pub available: bool,
    pub command: Option<String>,
    pub source: Option<String>,
}

#[tauri::command]
pub fn start_dap<R: Runtime>(
    app: AppHandle<R>,
    state: tauri::State<'_, DapState>,
    cmd: String,
    args: Vec<String>,
    cwd: Option<String>,
) -> Result<(), String> {
    if let Some(mut old_child) = state.child.lock().unwrap().take() {
        let _ = old_child.kill();
    }
    *state.stdin.lock().unwrap() = None;

    let mut command = Command::new(&cmd);
    command
        .args(&args)
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped());
    if let Some(cwd) = cwd.as_ref() {
        command.current_dir(cwd);
    }

    let mut child = command
        .spawn()
        .map_err(|e| format!("Failed to start DAP adapter '{}': {}", cmd, e))?;

    let stdout = child.stdout.take().ok_or("Failed to open DAP stdout")?;
    let stderr = child.stderr.take();
    let stdin = child.stdin.take().ok_or("Failed to open DAP stdin")?;

    *state.stdin.lock().unwrap() = Some(stdin);
    *state.child.lock().unwrap() = Some(child);

    let stdout_app = app.clone();
    thread::spawn(move || {
        let mut reader = BufReader::new(stdout);
        loop {
            let Some(len) = read_content_length(&mut reader) else {
                break;
            };

            let mut body = vec![0u8; len];
            if reader.read_exact(&mut body).is_err() {
                break;
            }

            let msg = String::from_utf8_lossy(&body).to_string();
            let _ = stdout_app.emit("dap-message", msg);
        }
    });

    if let Some(stderr) = stderr {
        let stderr_app = app.clone();
        thread::spawn(move || {
            let reader = BufReader::new(stderr);
            for line in reader.lines().map_while(Result::ok) {
                let _ = stderr_app.emit("dap-output", line);
            }
        });
    }

    Ok(())
}

#[tauri::command]
pub fn stop_dap(state: tauri::State<'_, DapState>) -> Result<(), String> {
    *state.stdin.lock().unwrap() = None;
    if let Some(mut child) = state.child.lock().unwrap().take() {
        let _ = child.kill();
        let _ = child.wait();
    }
    Ok(())
}

#[tauri::command]
pub fn send_dap_message(state: tauri::State<'_, DapState>, message: String) -> Result<(), String> {
    if let Some(stdin) = state.stdin.lock().unwrap().as_mut() {
        let full_msg = format!(
            "Content-Length: {}\r\n\r\n{}",
            message.as_bytes().len(),
            message
        );
        stdin
            .write_all(full_msg.as_bytes())
            .map_err(|e| e.to_string())?;
        stdin.flush().map_err(|e| e.to_string())?;
    }
    Ok(())
}

#[tauri::command]
pub fn resolve_dap_command(
    candidates: Vec<DapCommandCandidate>,
    cwd: Option<String>,
) -> Result<Option<ResolvedDapCommand>, String> {
    Ok(resolve_dap_command_inner(&candidates, cwd.as_deref()))
}

#[tauri::command]
pub fn check_dap_adapters(
    adapters: Vec<DapAdapterProbe>,
    cwd: Option<String>,
) -> Result<Vec<DapAdapterStatus>, String> {
    Ok(adapters
        .into_iter()
        .map(|adapter| {
            let resolved = resolve_dap_command_inner(&adapter.candidates, cwd.as_deref());
            DapAdapterStatus {
                id: adapter.id,
                label: adapter.label,
                languages: adapter.languages,
                available: resolved.is_some(),
                command: resolved.as_ref().map(|cmd| {
                    std::iter::once(cmd.cmd.clone())
                        .chain(cmd.args.iter().cloned())
                        .collect::<Vec<_>>()
                        .join(" ")
                }),
                source: resolved.map(|cmd| cmd.source),
            }
        })
        .collect())
}

fn read_content_length<R: BufRead>(reader: &mut R) -> Option<usize> {
    let mut content_length = None;

    loop {
        let mut line = String::new();
        match reader.read_line(&mut line) {
            Ok(0) | Err(_) => return None,
            Ok(_) => {}
        }

        let trimmed = line.trim_end_matches(['\r', '\n']);
        if trimmed.is_empty() {
            return content_length;
        }

        if let Some((name, value)) = trimmed.split_once(':') {
            if name.eq_ignore_ascii_case("content-length") {
                content_length = value.trim().parse::<usize>().ok();
            }
        }
    }
}

fn resolve_dap_command_inner(
    candidates: &[DapCommandCandidate],
    cwd: Option<&str>,
) -> Option<ResolvedDapCommand> {
    for candidate in candidates {
        if let Some(path) = resolve_command_path(&candidate.cmd, cwd) {
            let Some(args) = expand_command_args(&candidate.args, cwd) else {
                continue;
            };
            if !args_are_available(&args) {
                continue;
            }
            return Some(ResolvedDapCommand {
                cmd: path,
                args,
                source: candidate.cmd.clone(),
            });
        }
    }
    None
}

fn expand_command_args(args: &[String], cwd: Option<&str>) -> Option<Vec<String>> {
    let app_root = app_root().map(|path| path.to_string_lossy().to_string());
    args.iter()
        .map(|arg| {
            let mut next = arg.clone();
            if next.contains("{AIDA_APP_ROOT}") {
                let root = app_root.as_ref()?;
                next = next.replace("{AIDA_APP_ROOT}", root);
            }
            if next.contains("{WORKSPACE}") {
                let workspace = cwd?;
                next = next.replace("{WORKSPACE}", workspace);
            }
            Some(next)
        })
        .collect()
}

fn args_are_available(args: &[String]) -> bool {
    args.iter().all(|arg| {
        let path = Path::new(arg);
        if !path.is_absolute() {
            return true;
        }
        let Some(ext) = path.extension().and_then(|ext| ext.to_str()) else {
            return true;
        };
        !matches!(ext, "js" | "mjs" | "cjs") || path.exists()
    })
}

fn app_root() -> Option<PathBuf> {
    let mut dir = env::current_dir().ok()?;
    loop {
        if dir
            .join("scripts")
            .join("aida-node-dap-adapter.mjs")
            .exists()
        {
            return Some(dir);
        }
        if !dir.pop() {
            return None;
        }
    }
}

fn resolve_command_path(cmd: &str, cwd: Option<&str>) -> Option<String> {
    let direct = PathBuf::from(cmd);
    if direct.is_absolute() && direct.exists() {
        return Some(direct.to_string_lossy().to_string());
    }

    if let Some(cwd) = cwd {
        if let Some(path) = resolve_from_node_bin(Path::new(cwd), cmd) {
            return Some(path);
        }
    }

    if let Ok(current_dir) = env::current_dir() {
        if let Some(path) = resolve_from_node_bin(&current_dir, cmd) {
            return Some(path);
        }
    }

    let path_var = env::var_os("PATH")?;
    for dir in env::split_paths(&path_var) {
        for candidate in command_variants(cmd) {
            let path = dir.join(&candidate);
            if path.exists() {
                return Some(path.to_string_lossy().to_string());
            }
        }
    }
    None
}

fn resolve_from_node_bin(base: &Path, cmd: &str) -> Option<String> {
    let bin_dir = base.join("node_modules").join(".bin");
    for candidate in command_variants(cmd) {
        let path = bin_dir.join(&candidate);
        if path.exists() {
            return Some(path.to_string_lossy().to_string());
        }
    }
    None
}

fn command_variants(cmd: &str) -> Vec<String> {
    let path = Path::new(cmd);
    if path.extension().is_some() {
        return vec![cmd.to_string()];
    }

    if cfg!(windows) {
        return [".cmd", ".exe", ".bat", ""]
            .iter()
            .map(|ext| format!("{cmd}{ext}"))
            .collect();
    }

    vec![cmd.to_string()]
}

File: electron/ipc/dap.ts
import { ipcMain } from "electron";
import { spawn, type ChildProcessWithoutNullStreams } from "node:child_process";
import { sendToRenderer, type WindowProvider } from "./events";
import {
  normalizeSpawnCommand,
  readProtocolMessages,
  resolveCommand,
  type CommandCandidate,
  type ResolvedCommand,
} from "./process-utils";

type DapAdapterProbe = {
  id: string;
  label: string;
  languages: string[];
  candidates: CommandCandidate[];
};

type DapAdapterStatus = {
  id: string;
  label: string;
  languages: string[];
  available: boolean;
  command: string | null;
  source: string | null;
};

let child: ChildProcessWithoutNullStreams | null = null;

function frameMessage(message: string): string {
  return `Content-Length: ${Buffer.byteLength(message, "utf8")}\r\n\r\n${message}`;
}

function stopCurrent(): void {
  if (!child) return;
  child.kill();
  child = null;
}

export function registerDapHandlers(getWindow: WindowProvider): void {
  ipcMain.handle("start_dap", async (_event, args: {
    cmd?: string;
    command?: string;
    args?: string[];
    cwd?: string | null;
  }) => {
    stopCurrent();

    const requestedCommand = args.cmd ?? args.command;
    if (!requestedCommand) throw new Error("DAP command is required");

    const normalized = normalizeSpawnCommand(requestedCommand, args.args ?? []);
    child = spawn(normalized.command, normalized.args, {
      cwd: args.cwd ?? undefined,
      stdio: ["pipe", "pipe", "pipe"],
      windowsHide: true,
    });

    child.on("exit", () => {
      child = null;
    });
    readProtocolMessages(child.stdout, (message) => sendToRenderer(getWindow, "dap:message", message));
    child.stderr.on("data", (chunk: Buffer) => {
      sendToRenderer(getWindow, "dap:output", chunk.toString("utf8"));
    });
  });

  ipcMain.handle("stop_dap", async () => {
    stopCurrent();
  });

  ipcMain.handle("send_dap_message", async (_event, args: { message: string }) => {
    if (!child?.stdin.writable) return;
    child.stdin.write(frameMessage(args.message));
  });

  ipcMain.handle("resolve_dap_command", async (_event, args: {
    candidates?: CommandCandidate[];
    cwd?: string | null;
    adapter_type?: string;
  }): Promise<ResolvedCommand | null> => {
    return resolveCommand(args.candidates ?? candidatesForAdapter(args.adapter_type), args.cwd);
  });

  ipcMain.handle("check_dap_adapters", async (_event, args: {
    adapters?: DapAdapterProbe[];
    cwd?: string | null;
  }): Promise<DapAdapterStatus[]> => {
    return (args.adapters ?? defaultAdapterProbes()).map((adapter) => {
      const resolved = resolveCommand(adapter.candidates, args.cwd);
      return {
        id: adapter.id,
        label: adapter.label,
        languages: adapter.languages,
        available: Boolean(resolved),
        command: resolved ? [resolved.cmd, ...resolved.args].join(" ") : null,
        source: resolved?.source ?? null,
      };
    });
  });
}

function candidatesForAdapter(adapterType?: string): CommandCandidate[] {
  switch (adapterType) {
    case "aida-node": return [{ cmd: "node", args: ["{AIDA_APP_ROOT}/scripts/aida-node-dap-adapter.mjs"] }];
    case "python": return [{ cmd: "debugpy-adapter", args: [] }, { cmd: "python", args: ["-m", "debugpy.adapter"] }];
    case "go": return [{ cmd: "dlv", args: ["dap"] }];
    default: return [];
  }
}

function defaultAdapterProbes(): DapAdapterProbe[] {
  return [
    { id: "aida-node", label: "Aida Node / JavaScript", languages: ["js", "mjs", "cjs", "ts"], candidates: candidatesForAdapter("aida-node") },
    { id: "python", label: "Python / debugpy", languages: ["py"], candidates: candidatesForAdapter("python") },
    { id: "go", label: "Go / Delve", languages: ["go"], candidates: candidatesForAdapter("go") },
  ];
}
